queries/stack-graphs.tsg

```diff
@@ -0,0 +1,818 @@
+[
+  (associated_type type_parameters: (_) @child) @parent
+  (associated_type bounds: (_) @child) @parent
+  (enum_item type_parameters: (_) @child) @parent
+  (enum_item (where_clause) @child) @parent
+  (enum_item body: (_) @child) @parent
+  (function_item type_parameters: (_) @child) @parent
+  (function_item parameters: (_) @child) @parent
+  (function_item return_type: (_) @child) @parent
+  (function_item (where_clause) @child) @parent
+  (function_signature_item (_) @child) @parent
+  (impl_item type_parameters: (_) @child) @parent
+  (impl_item trait: (_) @child) @parent
+  (impl_item type: (_) @child) @parent
+  (impl_item (where_clause) @child) @parent
+  (impl_item body: (_) @child) @parent
+  (struct_item type_parameters: (_) @child) @parent
+  (struct_item (where_clause) @child) @parent
+  (struct_item body: (_) @child) @parent
+  (trait_item type_parameters: (_) @child) @parent
+  (trait_item bounds: (_) @child) @parent
+  (trait_item (where_clause) @child) @parent
+  (trait_item body: (_) @child) @parent
+  (type_item type_parameters: (_) @child) @parent
+  (type_item type: (_) @child) @parent
+  (union_item type_parameters: (_) @child) @parent
+  (union_item body: (_) @child) @parent
+
+  (function_type trait: (_) @child) @parent  ;; only for Fn, FnMut, FnOnce
+  (function_type parameters: (_) @child) @parent
+  (function_type return_type: (_) @child) @parent
+] {
+  let @child.scope = @parent.inner_scope
+}
+
+;; Block scopes.
+;;
+;; Items in a block are visible throughout the block. Let-declaration bindings
+;; in a block are visible starting with the next statement or expression.
+
+;; Create the item scope.
+(block) @block {
+  ;; TODO: Figure out how to create these nodes conditionally. They're only
+  ;; needed if the block contains items; otherwise these can just be aliases.
+  ;; The reason it isn't done is I don't know how to use the query language
+  ;; to distinguish blocks that have items from those that don't.
+  node @block.item_scope
+  edge @block.item_scope -> @block.scope
+
+  node @block.item_types
+  attr (@block.item_types) pop_symbol = "%type"
+  edge @block.item_scope -> @block.item_types
+  attr (@block.item_scope -> @block.item_types) precedence = 1
+}
+
+;; The next two rules define each child's `.block_scope_before`.
+(block . (_) @first_child) @block {
+  let @first_child.block_scope_before = @block.item_scope
+}
+
+(block (_) @prev . (_) @next) {
+  let @next.block_scope_before = @prev.block_scope_after
+}
+
+;; The next two rules define `.block_scope_after` for each statement child.
+(let_declaration pattern: (_) @pat) @stmt {
+  node @stmt.block_scope_after
+  edge @stmt.block_scope_after -> @pat.defs
+  attr (@stmt.block_scope_after -> @pat.defs) precedence = 1
+  edge @stmt.block_scope_after -> @stmt.block_scope_before
+}
+
+(block [
+  ;; This is every node type that can appear in a block as a statement, except
+  ;; `let_declaration`. Awkward but I don't know how else to write this.
```

Comment:
This is indeed hard to express neatly in tree-sitter queries. You could probably do away with the explicit list if you are a bit more liberal with the amount of graph nodes that you create. You could set up nodes for each statement, and chain them, and link up the `.defs` of each statement to the correct node in the chain. Most of these will just not point to anything useful.

Our aim is that nodes and edges are cheap, so cheap that you don't feel the need to bend the rules to minimize their use. If they are not cheap enough, it's definitely something we'd try to solve generally in the library.